### PR TITLE
[ENGA3-511]: Improve the implementation to prevent the customer entering redirect URL directly in the address bar of the browser

### PR DESCRIPTION
**Pull request by aashishgurung**
#### 1. Objective

Improve the implementation to prevent the customer entering redirect URL directly in the address bar of the browser

Jira Ticket: [#511](https://opn-ooo.atlassian.net/browse/ENGA3-511)

#### 2. Description of change

Previously, we relied on `Referer` header for Safari browser and `HTTP_SEC_FETCH_SITE` header for other browsers to prevent user from triggering redirect URL manually. We knew Referer is not reliable as some browsers limit access to not allow HTTP_REFERER to be passed but we choose it for Safari because i does not support `HTTP_SEC_FETCH_SITE`. 

This implementation was not 100% reliable as some payment gateway sets the value of `HTTP_SEC_FETCH_SITE` to `none` instead of `cross-site`. This will result in customer being redirected to abandoned cart page even though the transaction was legit.

With this PR, we removed the HEADERs checks. We create a random token, save it to the payment of the order and pass it as the query parameters with the return URI to the gateway. Once, it is redirected back, we check whether the token in the query parameter is same as the token saved in the payment of the concerned order. If yes then we proceed with the transaction else we redirect user to empty  card page with message `Invalid token. Please contact our support if you have any questions`.

Screen Shot 2565-10-11 at 14 06 05

 
#### 3. Quality assurance

- Test the cancel and success transaction with any payment methods.
  - Examples: 3DS and internet banking

- Try entering the redirect URL to the address bar of the browser. It should redirect back to empty car page with error message `Invalid token. Please contact our support if you have any questions`.
  - 3DS: http://localhost:8200/th/omise/callback/threedsecure/
  - Offsite: http://localhost:8200/th/omise/callback/offsite/
  
Link to screencast:  https://drive.google.com/file/d/1lmHnUb62TwWvZ33zpa8GRUDsxQmd9vGp/view?usp=sharing

**🔧 Environments:**

- Platform version: Magento 2.4.5
- Omise plugin version: Omise-Magento 2.29.1
- PHP version: 8.1

### CODE REVIEW

File: Controller/Callback/Offsite.php [hunk 1]
```diff
@@ -110,19 +113,30 @@ public function execute()
...
 
-        if ($order->getState() !== Order::STATE_PENDING_PAYMENT) {
-            $this->invalid($order, __('Invalid order status, cannot validate the payment. Please contact our
-            support if you have any questions.'));
+        $token = $this->request->getParam('token');
+
+        if (!$token || $payment->getAdditionalInformation('token') !== rtrim($token, "/")) {
+            $this->invalid(
+                $order,
+                __('Invalid token. Please contact our support if you have any questions.')
```
Comment (kan98) [inline Controller/Callback/Offsite.php:128]: I think this message is not too meaningful to the customer trying to pay. They may be confused by what token means. Can we say something like `Unknown error. Please contact our support if you have any questions`? also not as useful but won't make the customer confused what a token is used for

Comment (kan98) [reply Controller/Callback/Offsite.php:128]: or any other suggestions you might have 🙂

**Comment by som-m** [comment]
👍🏻 

image

**Review by tanawin-opn — APPROVED** [review]

File: Controller/Callback/Offsite.php (hunk 1 above)
Comment (aashishgurung) [reply Controller/Callback/Offsite.php:128]: Yeah you are right. `Invalid URL` might be better. I will ask the team for better suggestions. Thanks.

**Review by kan98 — APPROVED** [review]

**Pull request merged**